packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1257 @@
+  /// overrides this method to report the route information back to the engine.
+  ///
+  /// The [routeInformation] is the new route information after the navigation
+  /// event.
+  void routerReportsNewRouteInformation(RouteInformation routeInformation) {}
+}
+
+/// The route information provider that propagates the platform route information changes.
+///
+/// This provider also reports the new route information from the [Router] widget
+/// back to engine using message channel method, the
+/// [RouteNotificationMessages.maybeNotifyRouteChange].
+class PlatformRouteInformationProvider extends RouteInformationProvider with WidgetsBindingObserver, ChangeNotifier {
+  /// Create a platform route information provider.
+  ///
+  /// Use the [initialRouteInformation] to set the default route information for this
+  /// provider.
+  PlatformRouteInformationProvider({RouteInformation initialRouteInformation})
+      : _value = initialRouteInformation,
+        _currentRouteInformationInEngine = initialRouteInformation;
+
+  RouteInformation _currentRouteInformationInEngine;
+
+  @override
+  void routerReportsNewRouteInformation(RouteInformation routeInformation) {
+    RouteNotificationMessages.maybeNotifyRouteInformationChange(
+      routeInformation,
+      _currentRouteInformationInEngine
+    );
+    _currentRouteInformationInEngine = routeInformation;
+    _value = routeInformation;
+  }
+
+  void _updateAndNotify(RouteInformation newValue) {
+    assert(_value != newValue);
+    _value = newValue;
+    notifyListeners();
+  }
+
+  @override
+  RouteInformation get value => _value;
+  RouteInformation _value;
+  set value(RouteInformation newValue) {
+    if (_value == newValue)
+      return;
+    _updateAndNotify(newValue);
+    routerReportsNewRouteInformation(newValue);
+  }
+
+  void _platformReportsNewRouteInformation(RouteInformation routeInformation) {
+    if (_value == routeInformation)
+      return;
+    _updateAndNotify(routeInformation);
+    // We can't use the routerReportsNewRouteInformation because we don't want
+    // to report it back to the engine again.
+    _currentRouteInformationInEngine = routeInformation;
+  }
+
+  @override
+  void addListener(VoidCallback listener) {
+    if (!hasListeners)
+      WidgetsBinding.instance.addObserver(this);
+    super.addListener(listener);
+  }
+
+  @override
+  void removeListener(VoidCallback listener) {
+    super.removeListener(listener);
+    if (!hasListeners)
+      WidgetsBinding.instance.removeObserver(this);
+  }
+
+  @override
+  void dispose() {
+    // In practice, this will rarely be called. We assume that the listeners
+    // will be added and removed in a coherent fashion such that when the object
+    // is no longer being used, there's no listener, and so it will get garbage
+    // collected.
+    if (hasListeners)
+      WidgetsBinding.instance.removeObserver(this);
+    super.dispose();
+  }
+
```

Comment:
Should this also listen to didPushRoute?

Reply:
I remember we decided to make web engine notify both router and navigator channel. This will cause it to receive the event twice.

Reply:
The way it is it will stop working for all other embedders, though. Should we change it so that this listens to both and then document that embedder should call only one of them? We could also update WidgetApps to listen to the new one in addition to didPushRoute to make the old Navigator approach work for embedders that decide to call the new API?

Reply:
I am thinking maybe we should just make a breaking change to add a state parameter to didPushRoute and make both router and navigator channel uses it. The embedder can send message to whichever channel they like but just not both at the same time.

I can't think of other cleaner way to do this, either we need to deal with debouncing or the embedder need to migrate their code.

Reply:
Why wouldn't it work to have two methods and the embedder decides which one to call? The Router just responds to both.

Reply:
Existing apps will have to migrate their apps to listening on both, otherwise web application will break. Although they also need to migrate if we make the breaking change, the former will be a silent break which may be tricky to find out.

I think we will also want to merge these two if we start from scratch anyway. What do you think?

Reply:
Why do existing apps need to migrate? Isn't the change just to also listen to the other notification in the framework here:

https://github.com/flutter/flutter/blob/3aa0b48a2750197017c819588fa2e793244fa080/packages/flutter/lib/src/widgets/app.dart#L1058

Reply:
If we make a separate method for the route information channel. An app has a custom widgetbindingobserver will need to migrate their code to also listen to the new method. Otherwise, the widgetbindingobserver will not work properly in web application.

If we merge the two method, we will make a breaking change by introducing an additional state, and restoration state parameter in `didPushRoute`.

For both cases, existing app will need to migrate.

Reply:
Would it work to have two separate methods and the default implementation of WidgetsBindingObserver.didPushRouteInformation just calls WidgetsBindingObserver.didPush? That way, you can listen to them separately, but if you don't upgrade your app everything continues to work even when the engine calls the new methods.

Reply:
That is brilliant, I will do this way. Thanks!